codegen/type_converter.go

```diff
@@ -69,24 +69,101 @@ func (c *TypeConverter) getGoTypeName(valueType compile.TypeSpec) (string, error
 	return GoType(c.Helper, valueType)
 }
 
+//   input of "A.B.C.D" returns ["A","A.B", "A.B.C", "A.B.C.D"]
+func getMiddleIdentifiers(identifier string) []string {
+	subIds := strings.Split(identifier, ".")
+
+	middleIds := make([]string, 0, len(subIds))
+
+	middleIds = append(middleIds, subIds[0])
+	for i := 1; i < len(subIds); i++ {
+		middleIds = append(middleIds, fmt.Sprintf("%s.%s", middleIds[i-1], subIds[i]))
+	}
+
+	return middleIds
+}
+
+//  converts a list of identifier paths into boolean nil check expressions on those paths
+func convertIdentifiersToNilChecks(identifiers []string) []string {
+	checks := make([]string, 0, len(identifiers)-1)
+
+	for i, ident := range identifiers {
+		if i == 0 {
+			continue // discard the check on "in" object
+		}
+
+		checks = append(checks, ident+" != nil")
+	}
+	return checks
+}
+
+//  tranforms will have different paths,   to->from proxy mappings will have the same identifier path aside from prefix
+func isTransform(from string, to string) bool {
+	//  from prefix is "in." and to prefix "out."
+	return from[3:] != to[4:]
+}
+
+// helper func for assignWithOverride
+func (c *TypeConverter) assignWithChecks(
+	indent string,
+	toIdentifier string,
+	toTypeName string,
+	toRequired bool,
+	fromIdentifier string,
+	fromRequired bool,
+	extraNilCheck bool,
+) {
+
+	if !isTransform(fromIdentifier, toIdentifier) {
+
+		if !fromRequired && toRequired || extraNilCheck {
+			// need to nil check otherwise we could be cast or deref nil
+			c.append(indent, "if ", fromIdentifier, " != nil {")
+			c.append(indent, "\t", toIdentifier, " = ", toTypeName, "(", fromIdentifier, ")")
+			c.append(indent, "}")
+		} else {
+			c.append(indent, toIdentifier, " = ", toTypeName, "(", fromIdentifier, ")")
+		}
+		return
+	}
+
+	//  generates nil checks for intermediate objects on the fromIdentifier path
+	checks := convertIdentifiersToNilChecks(getMiddleIdentifiers(fromIdentifier))
+
+	if !extraNilCheck {
+		// if from is required or from/to are both optional   then we don't need the final outer check
+		if fromRequired || !fromRequired && !toRequired {
+			checks = checks[:len(checks)-1]
+		}
+		if len(checks) == 0 {
+			c.append(indent, toIdentifier, " = ", toTypeName, "(", fromIdentifier, ")")
+			return
+		}
+	}
+	c.append(indent, "if ", strings.Join(checks, " && "), " {")
+	c.append(indent, "\t", toIdentifier, " = ", toTypeName, "(", fromIdentifier, ")")
+	c.append(indent, "}")
+	// TODO else?  log/stat it?  should this siliently eat intermediate nils as none-assignement,  should set nil?
```

Comment:
I think optional/required validation is done somewhere at the thrift deserialize / serialize stage on both egress & ingress. 

So the proxy can be quite lax and just silently eat nils and not populate destination fields if its not possible to populate destination fields.

The proxy will do a best effort population of the downstream request and do "strict" validation of the inbound schema. It's the downstream servers responsibility to implement "strict" validation for their downstream server, so our proxy request to them can be "best effort" so that it works for "all opinions of strictness" for each possible downstream.